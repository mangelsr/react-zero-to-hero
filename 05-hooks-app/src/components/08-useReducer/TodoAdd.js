import React from 'react'
import { useForm } from '../../hooks/useForm';


export const TodoAdd = ({handleAddTodo}) => {

    const [ { desc }, handleInputChange, reset ] = useForm({ desc: '' });

    const handleSubmit = (e) => {
        e.preventDefault();

        if (desc.trim().length <= 1) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc,
            done: false
        }

        handleAddTodo(newTodo);

        reset();
    };

    return (
        <>
            <h4>Agregar TODO</h4>
            <hr/>
            <form onSubmit={handleSubmit}>
                <input 
                    type="text"
                    name="desc"
                    placeholder="Aprender..."
                    autoComplete="off"
                    className="form-control"
                    value={desc}
                    onChange={handleInputChange}
                />
                <button 
                    type="submit"
                    className="btn btn-outline-primary btn-block mt-2 ">
                    Agregar
                </button>
            </form>
        </>
    )
}
