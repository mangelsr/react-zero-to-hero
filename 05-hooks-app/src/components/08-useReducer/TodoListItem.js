import React from 'react'


export const TodoListItem = ({todo, i, handleDelete, handleToggle}) => {
    return (
        <li className="list-group-item">
            <p
                onClick={() => handleToggle(todo.id)}
                className={`${todo.done && 'done'}`}>
                {i+1}. {todo.desc}
            </p>
            <button
                onClick={() => handleDelete(todo.id)}
                className="btn btn-danger">
                Delete
            </button>
        </li>
    )
}
