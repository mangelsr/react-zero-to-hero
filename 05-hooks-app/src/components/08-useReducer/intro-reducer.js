console.log('Hola mundo');

const initialState = [{
    id: 1,
    todo: 'Comprar un PS4',
    done: false
}];

const todoReducer = (state = initialState, action) => {
    switch (action?.type) {
        case 'ADD':
            return [...state, action.payload];
        default:
            return state;
    }
}

let todos = todoReducer();

const newTodo = {
    id: 2,
    todo: 'Comprar KH3',
    done: false
};

const addTodoAction = {
    type: 'ADD', // standar
    payload: newTodo // standar para argumentos
};

todos = todoReducer(todos, addTodoAction);




console.log(todos);