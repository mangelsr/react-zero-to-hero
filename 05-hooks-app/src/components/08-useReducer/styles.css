p {
    margin: auto 0;
    cursor: pointer;
}

.list-group-item {
    display: flex;
    align-items: center;
    justify-content: space-between;
}

.done {
    text-decoration: line-through;
}