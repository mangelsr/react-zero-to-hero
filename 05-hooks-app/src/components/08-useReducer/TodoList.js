import React from 'react'

import { TodoListItem } from './TodoListItem'


export const TodoList = ({todos, handleDelete, handleToggle}) => {
    return (
        <ul className="list-group list-group-flush">
            {todos.map((todo, i) => 
                <TodoListItem
                    key={todo.id}
                    i={i}
                    todo={todo}
                    handleToggle={handleToggle}
                    handleDelete={handleDelete}
                />
            )}
        </ul>
    )
}
