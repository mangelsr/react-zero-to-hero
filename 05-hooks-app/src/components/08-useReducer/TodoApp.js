import React, { useEffect, useReducer } from 'react'

import { todoReducer } from './todoReducer';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';
import './styles.css';


const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
};


export const TodoApp = () => {

    const [ todos, dispatch ] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    const handleDelete = (todoId) => {
        const action = {
            type: 'DELETE',
            payload: todoId
        };
        dispatch(action);
    };

    const handleToggle = (todoId) => {
        const action = {
            type: 'TOGGLE',
            payload: todoId
        };
        dispatch(action);
    };

    const handleAddTodo = (newTodo) => {
        dispatch({
            type: 'ADD',
            payload: newTodo
        });
    };

    return (
        <div>
            <h1>Todo App ({todos.length})</h1>
            <hr/>
            <div className="row">
                <div className="col-7">
                    <TodoList
                        todos={todos}
                        handleToggle={handleToggle}
                        handleDelete={handleDelete}
                    />
                </div>
                <div className="col-5">
                    <TodoAdd handleAddTodo={handleAddTodo} />
                </div>
            </div>
        </div>
    )
}