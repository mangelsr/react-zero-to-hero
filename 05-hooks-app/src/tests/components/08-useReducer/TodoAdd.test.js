import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import React from 'react';
import { TodoAdd } from '../../../components/08-useReducer/TodoAdd';


describe('Tests TodoAdd component', () => {
    
    const handleAddTodo = jest.fn();
    const wrapper = shallow(<TodoAdd handleAddTodo={handleAddTodo} />);

    test('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should not call handleAddTodo', () => {
        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({ preventDefault(){} });
        expect(handleAddTodo).toHaveBeenCalledTimes(0);
    });

    test('should call handleAddTodo', () => {
        const value = 'Test todo';
        const input = wrapper.find('input');
        input.simulate('change', { target: { name: 'desc', value } });
        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({ preventDefault(){} });
        
        expect(handleAddTodo).toHaveBeenCalledTimes(1);
        
        expect(handleAddTodo).toHaveBeenCalledWith({
            id: expect.any(Number),
            desc: value,
            done: false
        });

        expect(input.prop('value')).toBe('');
    });
    
});