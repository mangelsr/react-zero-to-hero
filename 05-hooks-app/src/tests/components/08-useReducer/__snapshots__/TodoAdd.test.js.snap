// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests TodoAdd component should match snapshot 1`] = `
<Fragment>
  <h4>
    Agregar TODO
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="desc"
      onChange={[Function]}
      placeholder="Aprender..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary btn-block mt-2 "
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
